roles/StackStorm.nginx/tasks/nginx_redhat.yml

```diff
@@ -0,0 +1,123 @@
+---
+- name: Install vars
+  include_vars: "{{ item }}"
+  with_first_found:
+    - "{{ ansible_os_family | lower }}_{{ ansible_distribution_major_version }}.yml"
+    - "{{ ansible_os_family | lower }}.yml"
+
+- name: Add nginx key
+  become: yes
+  rpm_key:
+    key: http://nginx.org/keys/nginx_signing.key
+    state: present
+  register: _task
+  retries: 5
+  delay: 3
+  until: _task is succeeded
+  tags: nginx
+
+- name: Add nginx repos
+  become: yes
+  yum_repository:
+    name: nginx
+    description: nginx repo
+    baseurl: http://nginx.org/packages/rhel/{{ ansible_facts.distribution_major_version }}/x86_64/
+    gpgcheck: yes
+    enabled: yes
+    state: present
+  tags: nginx
+
+- name: Install nginx
+  become: yes
+  yum:
+    name: nginx
+    state: present
+    disablerepo: epel
+  register: _task
+  retries: 5
+  delay: 3
+  until: _task is succeeded
+  tags: nginx
+
+- name: Remove default site
+  become: yes
+  file:
+    path: /etc/nginx/conf.d/default.conf
+    state: absent
+  tags: nginx
+
+- name: Comment out server line
+  become: yes
+  replace:
+     path: /etc/nginx/nginx.conf
+     backup: yes
+     regexp: '^(?![#])(.*server\s*{)'
+     replace: '#\1'
+  tags: nginx
+  when: (ansible_facts.os_family == 'RedHat' and ansible_facts.distribution_major_version == '8')
+
+- name: Comment out after server block
+  become: yes
+  replace:
+     path: /etc/nginx/nginx.conf
+     backup: yes
+     after: '\s*server\s*{'
+     regexp: '^(?![#}])(.+)$'
+     replace: '#\1'
+  tags: nginx
+  when: (ansible_facts.os_family == 'RedHat' and ansible_facts.distribution_major_version == '8')
+
+- name: Install dependencies for SELinux Ansible module
+  become: yes
+  yum:
+    name: "{{ selinux_dependencies }}"
+    state: present
+  register: nginx_selinux_dependencies
+  retries: 5
+  delay: 3
+  until: nginx_selinux_dependencies is succeeded
+  tags: nginx
+
+- name: Update SELinux facts after installing dependencies
+  become: yes
+  setup:
+    filter: ansible_selinux
+  when: nginx_selinux_dependencies.changed
+  tags: nginx, skip_ansible_lint
+
+- name: Adjust SELinux to allow network access for nginx
+  become: yes
+  seboolean:
+    name: httpd_can_network_connect
+    state: yes
+    persistent: yes
+  when: ansible_facts.selinux.status == "enabled" and ansible_facts.selinux.mode == "enforcing"
+  tags: nginx
+
+- name: Discover if hypervisor is EC2
```

Comment:
OK, based on @amanda11 comment that we have some AWS EC2-specific rules to enable/disable firewall rules looks strange to me.

The original code based on @punkrokk initial work from the scripted installer:
https://github.com/StackStorm/st2-packages/blob/29a493e33de928e6df452a9102e80aaeadc5ec53/scripts/st2bootstrap-el8.sh#L685-L690

@punkrokk Can you shed more light about this case and why we don't want to change the firewall rules for AWS? How about other VMs, other cloud providers, bare-metal servers, etc? What's so specific about it?

Reply:
@armab @amanda11 Standard practice on AWS is to not run a firewall on the EC2 instance and rely on security groups. They are functionally the same thing, while security groups are much more flexible and centrally manageable. 

Most AMIs in EC2 ship without a host based firewall enabled. The reason this ended up in the Ansible playbook here is because I was testing on the official RHEL AMI, and this was a failure point in EC2. 

I think we may want to have other checks for GCP and Azure, but that should probably be a second ticket.

Reply:
OK, so was the original failure due to absense of the `firewall-cmd` binary or what?

Reply:
Yes it was caused by the absence of the firewall-cmd binary. And I don't believe we should install it.

Reply:
OK.
So sounds like original scripted installer instead of checking for `EC2` hypervisor, should check for `firewall-cmd` binary presence instead.

Reply:
I also agree we shouldn't try to install the binary.

@punkrokk Based on this discussion, can you please modify the original https://github.com/StackStorm/st2-packages/blob/29a493e33de928e6df452a9102e80aaeadc5ec53/scripts/st2bootstrap-el8.sh#L685-L690 scripted installer and check for `firewall-cmd` binary presence instead of ec2 to avoid such errors and confusion in future?

This will also make the rule work outside of EC2 in any provider which makes more sense.

Reply:
Maybe. But first, consider a situation where we desire to have firewall-cmd installed, outside of a cloud instance, and it isn't. How would we handle that?

Reply:
If there is a `firewall-cmd` present, - we always want to allow http and https to make st2web effectively work, doesn't matter cloud/VM/bare metal/etc.
If there is no `firewall-cmd` binary, - don't try to do anything.

What's the other case? I don't understand the statement about desire to have firewall-cmd installed. Can you clarify in more detail the scenario?

Reply:
So - I think this might open a can of worms. `firewalld` (like `iptables`) both are cli interfaces to the Linux Netfilter packet manipulation system. The best explanation of this I can find at the moment is: https://medium.com/@iced_burn/compare-firewalld-iptables-nftables-netfilter-de08a8d21b5b.

So - where I am stuck, is that just because `firewalld` or `iptables` is or isn't installed, that (in my mind) doesn't guarantee state. So, we could have a box that has iptables installed and a `deny all` on port 80 and 443, but everything will seem fine. 

At the end of the day, I think we all are in agreement that these ansible modules are desired to be applied to a standard (as in official ISO/AMI/etc) installation of a particular OS.

That said, we will and do have situations (testing this repo on AWS for example) that will necessitate certain things. 

I believe we can take the firewalld approach, but we should note that it's not as idempotent as we may believe, since the firewall can be modified with other tools (iptables/ufw/etc).

Reply:
FYI the approach we take in the Puppet module is that we do NOT manage the firewall. This is true for almost all puppet modules (few exceptions here and there).

The end-user can integrate then utilize our module to install the StackStorm bits that is responsible for, but leaves the rest of the system integration up to the end-user. 

The reasoning for this is similar to what @punkrokk said above. There are just too many variations with regards to firewall configurations, especially cross platform. From a Puppet perspective the only way you can guarantee configuration consistency is to have Puppet manage all rules in the firewall and purge (remove) any non-Puppet managed rules. Few organizations actually do this.

I would think of this similar to things like monitoring, authentication, centralized logging, etc. We don't and probably shouldn't worry about configuring these bits inside of our modules. Leave that to the end-user where those configurations are going to be unique to each end-user.

The only option/suggestion would be is to provide an "opt-in" flag where you can specify `manage_firewall` that would enable/disable firewall rule management and then a `firewall_type` or something along those lines. Depending on the `firewall_type` we could run the `firewall-cmd` parts. This way the user can choose the right option for them, and if they dont' like the way we configure the firewall, they can configure the firewall themselves.

Reply:
:+1: 
If user wants to manage the firewall rules, - they should use another Ansible role dedicated for that. This will also keep our roles cleaner.

Reply:
ok - so what approach did you agree to @armab ? :P

Reply:
> I also agree we shouldn't try to install the binary.
> 
> @punkrokk Based on this discussion, can you please modify the original https://github.com/StackStorm/st2-packages/blob/29a493e33de928e6df452a9102e80aaeadc5ec53/scripts/st2bootstrap-el8.sh#L685-L690 scripted installer and check for `firewall-cmd` binary presence instead of ec2 to avoid such errors and confusion in future?
> 
> This will also make the rule work outside of EC2 in any provider which makes more sense.

How do the more recent comments apply to your ask @armab

Reply:
@nmaludy had a good point. The same approach is followed by Chef. Ansible is no diff here with the separation of concerns via roles.

@punkrokk No firewall rule should be managed in nginx role so we'll just remove the `EC2 check` and `firewall-cmd` tasks alltogether. If user needs to manage firewall, - he should use dedicated role to do that, depending on his environment and toolset.

We also didn't add any firewall-management tasks before in nginx role.

Reply:
Ok. That works. I’d like to see the Readme update with a warning and a breaking change added to the change log.  I appreciate this direction and agree with it, but There will be some that miss this and will come for support.

Reply:
Can you clarify about the Readme update?
We didn't have firewall rules in Ansible roles before, where is the breaking change?

Reply:
So I will try and address the comments tomorrow, but it’s agreed that I shall just remove those firewall-cmd calls in the ansible playbook. 
I will address the other comment. I will take a look and see if I can find somewhere to update in docs. I presume we want to point out that on distributes where firewalld is enabled they will need to ensure that http and https are enabled.

Reply:
Cool :+1: 
And as a follow-up for scripted curl-bash installer: https://github.com/StackStorm/st2-packages/pull/650 so we have this logic improved based on discussion.
cc @punkrokk @nmaludy

Reply:
Maybe not a breaking change - but a warning about making sure you firewall is allowing traffic. You are right it’s not breaking